docs/code/MlNetCookBook.md

```diff
@@ -0,0 +1,707 @@
+var env = new LocalEnvironment();
+
+// Define the reader: specify the data columns and where to find them in the text file.
+var reader = TextLoader.CreateReader(env, ctx => (
+        // The four features of the Iris dataset will be grouped together as one Features column.
+        Features: ctx.LoadFloat(0, 3),
+        // Label: kind of iris.
+        Label: ctx.LoadText(4)
+    ),
+    // Default separator is tab, but the dataset has comma.
+    separator: ',');
+
+// Read the training data.
+var trainData = reader.Read(new MultiFileSource(dataPath));
+
+// Apply all kinds of standard ML.NET normalization to the raw features.
+var pipeline = reader.MakeNewEstimator()
+    .Append(r => (
+        MinMaxNormalized: r.Features.Normalize(fixZero: true),
+        MeanVarNormalized: r.Features.NormalizeByMeanVar(fixZero: false),
+        CdfNormalized: r.Features.NormalizeByCumulativeDistribution(),
+        BinNormalized: r.Features.NormalizeByBinning(maxBins: 256),
+    ));
+
+// Let's train our pipeline of normalizers, and then apply it to the same data.
+var normalizedData = pipeline.Fit(trainData).Transform(trainData);
+
+// Inspect one column of the resulting dataset.
+var meanVarValues = normalizedData.GetColumn(r => r.MeanVarNormalized).ToArray();
+```
+
+## How do I train my model on textual data?
+
+Generally speaking, *all ML.NET learners expect the features as a float vector*. So, if some of your data is not natively a float, you will need to convert to floats. 
+
+If we want to learn on textual data, we need to 'extract features' out of the texts. There is an entire research area of NLP (Natural Language Processing) that handles this. In ML.NET we offer some basic mechanisms of text feature extraction:
+- Text normalization (removing punctuation, diacritics, switching to lowercase etc.)
+- Separator-based tokenization.
+- Stopword removal.
+- Ngram and skip-gram extraction.
+- TF-IDF rescaling.
+- Bag of words conversion.
+
+ML.NET offers a "one-stop shop" operation called `TextFeaturizer`, that runs a combination of above steps as one big 'text featurization'. We have tested it extensively on text datasets, and we're confident that it performs reasonably well without the need to deep-dive into the operations. 
+
+However, we also offer a selection of elementary operations that let you customize your NLP processing. Here's the example below where we use them.
+
+Wikipedia detox dataset:
+```
+Sentiment   SentimentText
+1      Stop trolling, zapatancas, calling me a liar merely demonstartes that you arer Zapatancas. You may choose to chase every legitimate editor from this site and ignore me but I am an editor with a record that isnt 99% trolling and therefore my wishes are not to be completely ignored by a sockpuppet like yourself. The consensus is overwhelmingly against you and your trollin g lover Zapatancas,  
+1    ::::: Why are you threatening me? I'm not being disruptive, its you who is being disruptive.   
```

Comment:
>     [](start = 1, length = 4)

Your behavior with tabs is inconsistent in the examples. In the first example in this page, you used actual tabs, since the source file had tabs. In this example, the source file *also* has tabs, but we instead replaced them with spaces, but not spaces that follow any sort of regular pattern ... I see I think 7 spaces in the first, 4 in the second, 3 in the third, 5 in the fourth...